POLIO-1125 Multi Round Budget (Backend)

kemar opened the pull request:
Implement Polio Multi Round Budget.

Related JIRA tickets : [POLIO-1125](https://bluesquare.atlassian.net/browse/POLIO-1125)

## Important: this PR cannot be merged "as is" because it will break the front-end

I have listed below the API changes that the front-end must implement [to make this happen](https://whimsical.com/multi-round-budget-FdiBzSUsuozi8aAnuvJWAP).

## Changes

### 1) Models

<img width="1515" alt="before_after" src="https://github.com/BLSQ/iaso/assets/281139/4fda481d-de50-4a6e-a3bc-fece81639d2c">

Notable changes on `Campaign` model:

1. copied budget related fields to `BudgetProcess`
    - these fields must be removed from `Campaign` after this feature is proven successful
1. grouped a bunch of deprecated fields at the end of the model
    - we should clean these fields as they add layers of indirection

### 2) Data migration

New data migration to move data between `Campaign` and `BudgetProcess`.

This should be thoroughly reviewed as well (as the `BudgetProcess` model fields).

### 3) Admin

- new admin UI for `BudgetProcess`

### 4) Tests

- moved tests from `plugins.polio.budget.test` to `plugins.polio.api.budget`
- renamed the test class `TeamAPITestCase` to `BudgetCampaignViewSetTestCase`
- renamed 1 letter long variables ([see original file](https://github.com/BLSQ/iaso/blob/f968dc97d2b9c4a387afdc664ed99d402d32a2d6/plugins/polio/budget/test.py))
- added tests for previously untested things:
    - `CampaignBudgetSerializer.get_timeline()`
    - `TransitionOverrideSerializer`
        - fixed a permission bug making the path unreachable for non admin users
    - workflow categories
    - etc.

Note that there could still be untested areas of the code.

### 5) Serializers

`CampaignBudgetSerializer` no longer inherits from `CampaignSerializer` as it is now related to the `BudgetProcess`.

As a result, I deleted the following fields from `CampaignBudgetSerializer`:

- `cvdpv2_notified_at`
- `ordering_fields` (all were linked to `Campaign`)
- `filterset_fields` (all were linked to `Campaign`)

### 6) API

List of API changes to make it easier for the front-end.

_Not that things can be modified according to front-end needs_.

#### `POST /api/polio/budget/` (create a `BudgetProcess` instance)

POST data:

```
data={
    "rounds": [round_1.id, round_2.id],
},
```

Return:

```
{
    'id': 5,                // BudgetProcess ID.
    'created_by': {
        'first_name': 'test',
        'last_name': 'test',
        'username': 'test'
    },
    'created_at': '2024-02-08T15:39:55.648934Z',
    'rounds': [8, 9]
}
```

#### `GET /api/polio/budget/` (get a list of `BudgetProcess` instances)

Return before:

```
{
    'results': [{
        'created_at': '2024-02-08T15:22:24.328053Z',
        'id': '43e53995-cc54-4f2a-985c-f81d2b6d67a3', // Campaign ID.
        'obr_name': 'test campaign',
        'country_name': None,
        'current_state': {
            'key': '-',
            'label': '-'
        },
        'budget_last_updated_at': None
    }]
}
```

Return after:

```
{
    'results': [{
        'created_at': '2024-02-08T15:12:38.937758Z',
        'id': 17,                                   // BudgetProcess ID.
        'obr_name': 'test campaign',
        'country_name': 'ANGOLA',
        'round_numbers': [1, 2],                    // List of **numbers** of linked rounds.
        'current_state': {
            'key': '-',
            'label': '-'
        },
        'updated_at': '2024-02-08T15:12:38.937774Z'
    }]
}
```

#### `GET /api/polio/budget/id/` (get a `BudgetProcess` instance)

Before:

```
GET /api/polio/budget/campaign_id/
```

After:

```
GET /api/polio/budget/budget_process_id/
```

#### `POST /api/polio/budget/transition_to/` (make a `BudgetProcess` transition to the next step)

POST data before:

```
data={
    "transition_key": "submit_budget",
    "campaign": campaign.id,                 // Campaign ID.
    "comment": "comment",
    "files": [],
}
```

POST data after:

```
data={
    "transition_key": "submit_budget",
    "budget_process": budget_process.id,    // BudgetProcess ID.
    "comment": "comment",
    "files": [],
}
```

#### `POST /api/polio/budget/override/` (make a `BudgetProcess` transition to an arbitrary step)

POST data before:

```
{
    "new_state_key": "rejected",
    "campaign": campaign.id,                 // Campaign ID.
    "comment": "override me",
    "files": [],
}
```

POST data after:

```
{
    "new_state_key": "rejected",
    "budget_process": budget_process.id,    // BudgetProcess ID.
    "comment": "override me",
    "files": [],
}
```

## How to test

See unit tests.



[POLIO-1125]: https://bluesquare.atlassian.net/browse/POLIO-1125?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kemar committed:
Add a new Budget model

kemar committed:
Fix `test_step_files`

kemar committed:
Fix `test_transition_to_link`

kemar committed:
Fix `test_transition_to_link_json`

kemar committed:
Rename `Budget` to `BudgetProcess`

kemar committed:
Adapt `CampaignBudgetSerializer` and `TransitionToSerializer`

kemar committed:
Fix `ExportCampaignBudgetSerializer`

kemar committed:
Fix migrations

kemar committed:
Fix `BudgetStepQuerySet.filter_for_user`

kemar committed:
Fix `test_csv_export_date`

kemar committed:
Fix `BudgetStepQuerySet.filter_for_user`

kemar committed:
Group deprecated fields in `Campaign`

kemar committed:
Remove `ordering_fields` and `filterset_fields` from `BudgetCampaignViewSet`

kemar committed:
Rename test

kemar committed:
Adapt and fix `TransitionOverrideSerializer`

kemar committed:
Fix `test_simple_get_list_with_all_fields`

kemar committed:
Rebase migrations

kemar committed:
Test `CampaignBudgetSerializer.get_timeline()`

kemar committed:
Fix `test_simple_get_list_with_all_fields`

kemar committed:
Rename serializers

kemar committed:
Remove unused imports

kemar committed:
Fix migrations

kemar committed:
Fix and test `filter_queryset`

kemar committed:
Use `assertNumQueries`

kemar committed:
Move tests

kemar committed:
Rebase me

kemar committed:
Migrate data from `Campaign` to `Budget`

kemar committed:
Add `BudgetProcessAdmin`

kemar committed:
Ensure rounds are from the same campaign

kemar committed:
Update comment

madewulf commented on plugins/polio/budget/api.py:
```diff
@@ -52,50 +50,39 @@ class BudgetCampaignViewSet(ModelViewSet, CSVExportMixin):
 
+    def get_serializer_class(self):
+        if self.action == "create":
+            return BudgetProcessWriteSerializer
```
Just curious, is this a common thing to do? Having a different serializer for write operations?

kemar replied:
I don't really know if it's a common thing to do, but sometimes it makes things clearer IMO.

madewulf commented on plugins/polio/migrations/0167_migrate_budget_data.py:
```diff
@@ -0,0 +1,59 @@
+                approved_by_unicef_at_WFEDITABLE=campaign.approved_by_unicef_at_WFEDITABLE,
+                approved_at_WFEDITABLE=campaign.approved_at_WFEDITABLE,
+                approval_confirmed_at_WFEDITABLE=campaign.approval_confirmed_at_WFEDITABLE,
+                payment_mode=campaign.payment_mode or "",
```
why the or? Replacing None by blank spaces?

kemar replied:
Yes, because I used:

```python
payment_mode = models.CharField(max_length=30, choices=PAYMENT, blank=True)
```

The Django convention is to avoid using `null=True` for `CharField` and `TextField` since these fields could have two different values for “no data”: `None` or an empty string `""`.

https://sentry.io/answers/django-difference-between-null-and-blank/

madewulf commented on plugins/polio/migrations/0167_migrate_budget_data.py:
```diff
@@ -0,0 +1,59 @@
+# Generated by Django 4.2.9 on 2024-02-07 16:28
+
```
I doubt that Django generated that migration or it's way smarter than I thought. 

Joke aside, this migration is what I expected to see indeed.

madewulf reviewed (approved):
This looks good in terms of models and migrations. I guess that the work on frontend might still have impact on the APIs, but the most important thing, the DB structure, seems correct to me. Thanks!

kemar committed:
Fix migrations

kemar committed:
Fix migrations

kemar committed:
Display "Rounds" instead of "Virus notification date" and fix list params

kemar committed:
Fix `current_state` filter

kemar committed:
Fix `CustomFilterBackend` for `BudgetProcess`

kemar committed:
Use a proper FilterSet and rename front URL params
- `country__id__in` -> `countries`
- `current_state_key__in` -> `current_state_key`

kemar committed:
Search only in `obr` or `country` names and remove `CustomFilterBackend`

kemar committed:
Fix search by `org_unit_groups`

kemar committed:
Fix `get_next_transitions()`

kemar committed:
Use a new URL for the detail of a budget process

kemar committed:
Fix `/api/polio/budgetsteps/` to use `budget_process_id`

kemar committed:
Fix tests

kemar committed:
Fix docstring

kemar committed:
Display round numbers

kemar committed:
Fix `formatRoundNumbers`

kemar committed:
Fix TypeScript errors

kemar committed:
Add tests for `BudgetCampaignFilter`

kemar committed:
Test filtering on `BudgetStepViewSet`

kemar committed:
Fix migrations

kemar committed:
New endpoint to returns info about rounds for which a budget process can be created

kemar committed:
Fix query

kemar committed:
Remove custom hooks

kemar committed:
Trigger GET's `select` only once via `useCallback`

kemar committed:
Add translations

kemar committed:
Add yup validation schema

kemar committed:
Handle Budget Process creation

kemar committed:
Fix migrations and `on_delete` constraints

kemar committed:
Filter campaign without country

kemar committed:
Order `rounds__number`

kemar committed:
Use `on_delete=models.PROTECT`

kemar committed:
Implement `delete` on `BudgetCampaignViewSet`

kemar committed:
Allow to delete a BudgetProcess

kemar committed:
Add comments

kemar committed:
Fix translations

kemar committed:
Fix performance in `available_rounds_dropdowns`

kemar committed:
Make the API return `campaign_id`

kemar committed:
Add a `available_rounds` route

kemar committed:
Allow to edit a Budget Process (WIP)

kemar committed:
Use a nested serializer for Rounds instead of `ArrayAgg`

kemar committed:
Implement partial update on `BudgetCampaignViewSet`

kemar committed:
Implement partial update on the front-end

kemar committed:
Group similar API calls

kemar committed:
Fix yup's validation schemas

kemar committed:
Fix select callbacks

kemar committed:
Fix callbacks

kemar committed:
Factorize GET available rounds

kemar committed:
Fix migrations

kemar committed:
Export rounds in CSV

kemar committed:
Fix transition override

kemar committed:
Fix migrations

kemar committed:
Fix transition_to step

kemar committed:
Fix "Re-send step for corrections or as reminder"

kemar committed:
Fix round creation

kemar committed:
Prevent deletion of rounds linked to budget processes

kemar committed:
Rename `BudgetCampaign` to `BudgetProcess`

kemar committed:
Fix ordering

kemar committed:
Fix links in Polio emails and add tests

kemar committed:
Fix tests

kemar committed:
Fix typos

kemar committed:
Change wording

kemar committed:
Add rounds info in emails

kemar committed:
Replace Fragment with FormikProvider

kemar committed:
Replace Fragment with FormikProvider

kemar committed:
Remove `useCallback`

kemar committed:
Use `snackSuccessMessage`

kemar committed:
Update plugins/polio/js/src/domains/Budget/BudgetProcess/CreateBudgetProcessModal.tsx

Co-authored-by: Quang Son Le <38907762+quang-le@users.noreply.github.com>

kemar committed:
Update plugins/polio/js/src/domains/Budget/BudgetProcess/CreateBudgetProcessModal.tsx

Co-authored-by: Quang Son Le <38907762+quang-le@users.noreply.github.com>

kemar committed:
Fix migrations

kemar committed:
Add remaining budget fields to `BudgetProcess` read and write endpoints

kemar committed:
Test all fields in serializer tests

kemar committed:
Fix `test_create`

kemar committed:
Fix missing `request` object

kemar committed:
Add test for a set of fields only

kemar committed:
Fix tests

beygorghor committed:
moving slowly form

beygorghor committed:
working part 1

beygorghor committed:
saving

kemar committed:
Handle `BudgetProcess` updates and partial updates

beygorghor committed:
split form

beygorghor committed:
Merge branch 'POLIO-1237-multi-round-budget-tab' into POLIO-1237-multi-round-budget-tab-front

kemar committed:
Add tests for PATCH

beygorghor committed:
adding tab

beygorghor committed:
Merge branch 'POLIO-1237-multi-round-budget-tab' into POLIO-1237-multi-round-budget-tab-front

beygorghor committed:
edition done, just missing schema validation

beygorghor committed:
wording

beygorghor committed:
edit and create using serializer

beygorghor committed:
renam, fine tuning, change values on create too

beygorghor committed:
black and translations

beygorghor committed:
remove dead code

beygorghor committed:
fix test_budgets and validation

beygorghor committed:
fixing tests

beygorghor committed:
code review

kemar committed:
Merge pull request #1279 from BLSQ/POLIO-1237-multi-round-budget-tab-front

POLIO-1237 [front-end] multi round budget tab

kemar committed:
Merge pull request #1259 from BLSQ/POLIO-1237-multi-round-budget-tab

POLIO-1237 Move budget from campaign modal to budget process page

kemar committed:
Merge pull request #1166 from BLSQ/POLIO-1126-multi-round-budget-front

POLIO-1126 Multi Round Budget (Frontend)
